packages/lib-core/src/runtime/PluginLoader.ts

```diff
@@ -0,0 +1,553 @@
+ */
+export class PluginLoader {
+  private readonly options: Required<PluginLoaderOptions>;
+
+  /** Plugins processed by this loader. */
+  private readonly plugins = new Map<string, PluginLoadData>();
+
+  /** Subscribed event listeners. */
+  private readonly listeners = new Set<PluginLoadListener>();
+
+  constructor(options: PluginLoaderOptions = {}) {
+    this.options = {
+      canLoadPlugin: options.canLoadPlugin ?? (() => true),
+      canReloadScript: options.canReloadScript ?? (() => true),
+      fetchImpl: options.fetchImpl ?? basicFetch,
+      fixedPluginDependencyResolutions: options.fixedPluginDependencyResolutions ?? {},
+      sharedScope: options.sharedScope ?? {},
+      postProcessManifest: options.postProcessManifest ?? _.identity,
+      entryCallbackName: options.entryCallbackName ?? DEFAULT_REMOTE_ENTRY_CALLBACK,
+      getPluginEntryModule: options.getPluginEntryModule ?? _.noop,
+    };
+  }
+
+  /**
+   * Subscribe to plugin load events.
+   *
+   * Returns a function for unsubscribing the provided listener.
+   */
+  subscribe(listener: PluginLoadListener): VoidFunction {
+    this.listeners.add(listener);
+
+    let isSubscribed = true;
+
+    return () => {
+      if (isSubscribed) {
+        isSubscribed = false;
+
+        this.listeners.delete(listener);
+      }
+    };
+  }
+
+  private invokeListeners(...args: Parameters<PluginLoadListener>) {
+    this.listeners.forEach((listener) => {
+      listener(...args);
+    });
+  }
+
+  /**
+   * Load a plugin from the given URL.
+   *
+   * This involves the following asynchronous operations:
+   * - load plugin manifest (unless provided as an object)
+   * - resolve plugin dependencies
+   * - load plugin scripts
+   *
+   * Plugins using the `callback` registration method are expected to call the global entry
+   * callback function created via {@link registerPluginEntryCallback} method, passing two
+   * arguments: plugin name and the entry module.
+   *
+   * For plugins using the `custom` registration method, the `getPluginEntryModule` function
+   * is expected to return the entry module of the given plugin. If not implemented properly,
+   * plugins using the `custom` registration method will fail to load.
+   *
+   * Use `subscribe` method to respond to plugin load results.
+   */
+  async loadPlugin(
+    baseURL: string,
+    manifestNameOrObject: string | PluginManifest = DEFAULT_PLUGIN_MANIFEST,
+  ) {
+    let manifest: PluginManifest;
+
+    try {
+      manifest =
+        typeof manifestNameOrObject === 'string'
+          ? await this.loadPluginManifest(resolveURL(baseURL, manifestNameOrObject))
+          : manifestNameOrObject;
+    } catch (e) {
+      this.invokeListeners({
+        success: false,
+        errorMessage: 'Failed to load plugin manifest',
+        errorCause: e,
+      });
+      return;
+    }
+
+    try {
+      manifest = pluginManifestSchema.strict(true).validateSync(manifest, { abortEarly: false });
```

Comment:
We don't validate the generated plugin manifest, because we're the ones generating it here :smiley: 

We originally validated only some `DynamicRemotePlugin` constructor options (`pluginMetadata` and `extensions`). Current code validates the whole "options" object passed to `DynamicRemotePlugin` constructor.